### PR TITLE
feat(conventional-commits): Add conventional prerelease/graduation

### LINKED ISSUE #1433: Correct way to use conventional-commits and canary / prerelease
## Question
Made the switch to use conventional-commits and am trying to work out the "correct" way to do a beta or rc release.  My instinct was to do a:
```
lerna publish --exact --conventional-commits  --preid=beta
```
Or
```
lerna publish --exact --conventional-commits  --canary=beta
```
But neither appear to do what I would expect?

### LINKED ISSUE #1675: Graduating a prerelease with no further changes
When we're happy with our latest prerelease we'd like to graduate the packages even if no further changes were made.
We thought this would be fixed in issue #1671, but we now get a new problem where lerna wont release an already released commit. 

## Expected Behavior
Version is updated from 0.0.1-prelease.0 to 0.0.1

## Current Behavior
```
lerna publish --force-publish=lobby --yes --conventional-commits --no-verify-registry --no-verify-access
lerna notice cli v3.3.2
lerna info versioning independent
lerna info Looking for changed packages since lobby@0.0.1-prerelease.0
lerna WARN force-publish lobby
lerna ERR! ERELEASED The current commit has already been released. Please make new commits before continuing.
```

## Steps to Reproduce (for bugs)
1. `git commit -m'chore: example change'`
2. `lerna version prerelease --force-publish=lobby --yes --preid=prerelease --npm-tag=prerelease` => `0.0.1-prerelease.0`
3. `lerna version --force-publish=lobby --yes --conventional-commits` => would like `0.0.1` got `The current commit has already been released`

<details><summary>lerna.json</summary><p>

```json
{
  "packages": [
    "packages/*",
    "apps/*"
  ],
  "command": {
    "publish": {
      "conventionalCommits": true,
      "message": "chore(release): publish",
      "allowBranch": "release/build"
    }
  },
  "npmClient": "yarn",
  "useWorkspaces": true,
  "version": "independent"
}
```
</p></details>

## Context
We prerelease packages on every merge to our release branch. After testing we will manually take the decision to graduate these packages.

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.3.2 |
| `npm --version`  | 6.4.1 |
| `yarn --version` | 1.9.4 |
| `node --version` | v8.11.1 |

| OS | Version |
| --- | --- |
| macOS el Capitan | 10.11.6 |

### PR DESCRIPTION
**Pull request by erquhart**
## Description

Provides `--conventional-commits` recommended bumps for prerelease packages. Ensures that the current prerelease version, as well as the eventual graduated version, are semver compliant based on the commit messages of contained changes. When additional prereleases occur and the current prerelease base version is already adequate, the numeric prerelease identifier is incremented.

Prerelease bump logic is best illustrated in related tests:

https://github.com/lerna/lerna/blob/72f922ae9b9ce0a87035a80b1bda6f05dc703183/core/conventional-commits/__tests__/conventional-commits.test.js#L262-L285

Updates `--conventional-commits` to keep prereleased packages in prerelease and stable packages out of prerelease, but still providing semver bumps as expected.

#### Flags for use with `--conventional-commits`:

**`--conventional-prerelease`**
Releases all unreleased changes to pre(patch/minor/major/release) by prefixing the recommendation from `conventional-changelog`. Optionally accepts specific packages, in which case unspecified packages with changes present will be processed as they normally would using `--conventional-commits`.

**`--conventional-graduate`**
Graduates all prereleased packages, whether the current HEAD has been released or not. Similar to `--force-publish`, but ignores non-prerelease packages (unless they depend on a package that is being graduated). If unreleased changes are present, and the base version is inadequate for those changes, eg. features in a prepatch, the resulting package version(s) will reflect the recommended bump from `conventional-commits`. Optionally accepts specific packages, in which case unspecified packages will be processed as they normally would using `--conventional-commits`.

#### Flags for use with or without `--conventional-commits`:

**`--pre-dist-tag`**
Works identically to `--dist-tag`, except only applies for packages being released with a prerelease version. Necessary for proper npm dist tagging with independent versioning, where prerelease and stable versions may be released at once, and require different tags. Can be combined with `--dist-tag`.

## Motivation and Context

Fixes #1433 by providing a predictable, CI compatible approach to prerelease publishing for projects using conventional commits, including those using independent or fixed versioning. My immediate context is a desire to publish `beta` releases from `master` every time a new PR is merged.

Fixes #1675 by using `checkWorkingTree.uncommittedChanges` directly in the `version` command script when `--force-publish` (or `--conventional-graduate`) is set.

## How Has This Been Tested?

- Covered all changes with integration and unit tests.
- Running production prereleases/releases of [`netlify-cms`](https://npmjs.com/package/netlify-cms) using a published testing fork of this PR ([`@erquhart/lerna`](https://npmjs.com/package/@erquhart/lerna)).

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by pantoninho** [comment]
You sir, have made my day. Can’t wait to try this out.

Wonderful job.

File: commands/version/index.js [hunk 1]
```diff
@@ -346,6 +377,7 @@ class VersionCommand extends Command {
           changelogPreset,
           rootPath,
           tagPrefix: this.tagPrefix,
+          prereleaseId: getPrereleaseId(node),
```
Comment (erquhart) [inline commands/version/index.js]: Logic determining whether to prerelease stable packages (`shouldPrerelease`) or graduate prereleasePackages (`shouldGraduate`) is encapsulated here so that `ConventionalCommitUtilties.recommendVersion` is only reacting to the presence or absence of `prereleaseId`.

File: core/conventional-commits/lib/recommend-version.js [hunk 2]
```diff
@@ -35,8 +49,15 @@ function recommendVersion(pkg, type, { changelogPreset, rootPath, tagPrefix }) {
         // we still need to bump _something_ because lerna saw a change here
         const releaseType = data.releaseType || "patch";
 
-        log.verbose(type, "increment %s by %s", pkg.version, releaseType);
-        resolve(semver.inc(pkg.version, releaseType));
+        if (prereleaseId) {
```
Comment (erquhart) [inline core/conventional-commits/lib/recommend-version.js]: `prereleaseId` both activates prerelease recommendations and provides the prerelease ID to be used.

File: utils/collect-updates/collect-updates.js [hunk 3]
```diff
@@ -43,45 +48,36 @@ function collectUpdates(filteredPackages, packageGraph, execOpts, commandOptions
...
     // "warn" might seem a bit loud, but it is appropriate for logging anything _forced_
-    log.warn("force-publish", forced.has("*") ? "all packages" : Array.from(forced.values()).join("\n"));
+    const forceLabel = useConventionalGraduate ? "conventional-graduate" : "force-publish";
+    log.warn(forceLabel, forced.has("*") ? "all packages" : Array.from(forced.values()).join("\n"));
   }
 
-  let candidates;
+  // --conventional-commits --conventional-graduate
+  if (useConventionalGraduate) {
+    if (forced.has("*")) {
```
Comment (erquhart) [inline utils/collect-updates/collect-updates.js]: If `--conventional-graduate` is different from `--force-publish` in that it only impacts packages that are in prerelease and never affects packages that are not in any way. This is why we don't set all packages as candidates here as we do with `--force-publish`.

File: utils/collect-updates/lib/collect-packages.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+"use strict";
+
+const collectDependents = require("./collect-dependents");
+
+module.exports = collectPackages;
+
+function collectPackages(packages, { isCandidate = () => true, onInclude } = {}) {
```
Comment (erquhart) [inline utils/collect-updates/lib/collect-packages.js]: This felt like a smart abstraction as prerelease functionality introduced a need to collect dependents from outside of `collect-updates`. Tests were added for the new function.

File: utils/collect-updates/lib/get-packages-for-option.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+"use strict";
+
+module.exports = getPackagesForOption;
+
+function getPackagesForOption(option) {
```
Comment (erquhart) [inline utils/collect-updates/lib/get-packages-for-option.js]: Changed `getForcedPackages` to this more generic function for use with options besided `--force-publish`.

File: commands/version/index.js [hunk 6]
```diff
@@ -232,7 +243,10 @@ class VersionCommand extends Command {
 
     // amending a commit probably means the working tree is dirty
     if (this.commitAndTag && this.gitOpts.amend !== true) {
-      tasks.unshift(() => checkWorkingTree(this.execOpts));
+      const { forcePublish, conventionalCommits, conventionalGraduate } = this.options;
+      const checkUncommittedOnly = forcePublish || (conventionalCommits && conventionalGraduate);
+      const check = checkUncommittedOnly ? checkWorkingTree.throwIfUncommitted : checkWorkingTree;
```
Comment (erquhart) [inline commands/version/index.js]: These three lines are the fix for #1675.

File: commands/version/index.js [hunk 7]
```diff
@@ -232,7 +243,10 @@ class VersionCommand extends Command {
...
+      tasks.unshift(() => check(this.execOpts));
```
Comment (erquhart) [inline commands/version/index.js]: These four lines are the fix for #1675.

**Comment by erquhart** [comment]
`--conventional-graduate` in action with a released `HEAD` (okay there was one commit added but it was a chore that didn't touch these packages!):

```
$ run-s "publish:version --conventional-graduate"
$ lerna version --no-push --conventional-graduate
lerna notice cli v3.14.1
lerna WARN conventional-graduate all packages

Changes:
 - netlify-cms-backend-bitbucket: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-backend-git-gateway: 2.3.0-beta.0 => 2.3.0
 - netlify-cms-backend-github: 2.3.0-beta.0 => 2.3.0
 - netlify-cms-backend-gitlab: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-backend-test: 2.1.0-beta.0 => 2.1.0
 - netlify-cms-core: 2.9.0-beta.0 => 2.9.0
 - netlify-cms-editor-component-image: 2.3.0-beta.0 => 2.3.0
 - netlify-cms-lib-auth: 2.1.0-beta.0 => 2.1.0
 - netlify-cms-lib-util: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-media-library-cloudinary: 1.2.0-beta.0 => 1.2.0
 - netlify-cms-media-library-uploadcare: 0.4.0-beta.0 => 0.4.0
 - netlify-cms-ui-default: 2.5.0-beta.0 => 2.5.0
 - netlify-cms-widget-boolean: 2.1.0-beta.0 => 2.1.0
 - netlify-cms-widget-date: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-widget-datetime: 2.1.0-beta.0 => 2.1.0
 - netlify-cms-widget-file: 2.3.0-beta.0 => 2.3.0
 - netlify-cms-widget-image: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-widget-list: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-widget-map: 1.2.0-beta.0 => 1.2.0
 - netlify-cms-widget-markdown: 2.3.0-beta.0 => 2.3.0
 - netlify-cms-widget-number: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-widget-object: 2.1.0-beta.0 => 2.1.0
 - netlify-cms-widget-relation: 2.2.0-beta.0 => 2.2.0
 - netlify-cms-widget-select: 2.3.0-beta.0 => 2.3.0
 - netlify-cms-widget-string: 2.1.0-beta.0 => 2.1.0
 - netlify-cms-widget-text: 2.1.0-beta.0 => 2.1.0
 - netlify-cms: 2.7.0-beta.0 => 2.7.0

? Are you sure you want to create these versions? (ynH)
```

**Comment by evocateur** [comment]
@erquhart This is exciting, I will have time to review soon!

**Comment by erquhart** [comment]
Update: still working without issue. We now have a workflow of prereleasing to `@beta`, and always doing one final prerelease before graduating to stable, so that beta channel users always have the latest stable changes when available.

Recent dry run showing graduations after a few rounds of prerelease:

**Latest commit (Lerna release)**
```
$ git log -1
commit 002b8b43459a319f1ddff40816891aef272c7eed (HEAD -> master, tag: netlify-cms@2.9.0-beta.1, tag: netlify-cms-app@2.9.0-beta.1, upstream/master, upstream/H
Author: Shawn Erquhart <shawn@erquh.art>
Date:   Fri Mar 29 12:54:48 2019 -0400

    chore(release): publish

     - netlify-cms-app@2.9.0-beta.1
     - netlify-cms@2.9.0-beta.1
```

**Dry run of `lerna version --conventional-commits --conventional-graduate` with released head**
```
$ lerna version --no-push --conventional-graduate --no-git-tag-version
lerna notice cli v3.14.1
lerna WARN conventional-graduate all packages
lerna WARN version Skipping working tree validation, proceed at your own risk

Changes:
 - netlify-cms-app: 2.9.0-beta.1 => 2.9.0
 - netlify-cms-backend-bitbucket: 2.3.1-beta.1 => 2.3.1
 - netlify-cms-backend-git-gateway: 2.4.1-beta.1 => 2.4.1
 - netlify-cms-backend-github: 2.4.1-beta.1 => 2.4.1
 - netlify-cms-backend-gitlab: 2.3.1-beta.1 => 2.3.1
 - netlify-cms-backend-test: 2.2.1-beta.1 => 2.2.1
 - netlify-cms-core: 2.10.1-beta.2 => 2.10.1
 - netlify-cms-default-exports: 2.2.1-beta.2 => 2.2.1
 - netlify-cms-editor-component-image: 2.4.1-beta.1 => 2.4.1
 - netlify-cms-lib-auth: 2.2.1-beta.1 => 2.2.1
 - netlify-cms-lib-util: 2.3.1-beta.1 => 2.3.1
 - netlify-cms-media-library-cloudinary: 1.3.1-beta.1 => 1.3.1
 - netlify-cms-media-library-uploadcare: 0.5.1-beta.1 => 0.5.1
 - netlify-cms-ui-default: 2.6.1-beta.1 => 2.6.1
 - netlify-cms-widget-boolean: 2.2.1-beta.2 => 2.2.1
 - netlify-cms-widget-date: 2.3.1-beta.2 => 2.3.1
 - netlify-cms-widget-datetime: 2.2.1-beta.2 => 2.2.1
 - netlify-cms-widget-file: 2.4.1-beta.2 => 2.4.1
 - netlify-cms-widget-image: 2.3.1-beta.2 => 2.3.1
 - netlify-cms-widget-list: 2.3.1-beta.2 => 2.3.1
 - netlify-cms-widget-map: 1.3.1-beta.2 => 1.3.1
 - netlify-cms-widget-markdown: 2.4.1-beta.2 => 2.4.1
 - netlify-cms-widget-number: 2.3.1-beta.2 => 2.3.1
 - netlify-cms-widget-object: 2.2.1-beta.2 => 2.2.1
 - netlify-cms-widget-relation: 2.3.1-beta.2 => 2.3.1
 - netlify-cms-widget-select: 2.4.1-beta.2 => 2.4.1
 - netlify-cms-widget-string: 2.2.1-beta.2 => 2.2.1
 - netlify-cms-widget-text: 2.2.1-beta.2 => 2.2.1
 - netlify-cms: 2.9.0-beta.1 => 2.9.0

? Are you sure you want to create these versions? (ynH)
```

File: commands/publish/index.js [hunk 8]
```diff
@@ -685,7 +689,9 @@ class PublishCommand extends Command {
     };
     const mapper = pkg => {
       const spec = `${pkg.name}@${pkg.version}`;
-      const distTag = getDistTag(pkg.get("publishConfig"));
+      const preDistTag = this.getPreDistTag(pkg);
+      const distTag = preDistTag || getDistTag(pkg.get("publishConfig"));
+      opts.tag = preDistTag || opts.tag;
```
Comment (evocateur) [inline commands/publish/index.js]: We really shouldn't be mutating the outer `opts.tag` inside the mapper. The only value that matters here is `distTag`, the parameter passed to `npmDistTag.add()`, `opts.tag` _should_ never be considered when the underlying code executes (it's always overwritten by the parameter [here](https://github.com/lerna/lerna/blob/4e3f8d8087e0e2e8e83c7ea79558c517203583c2/utils/npm-dist-tag/npm-dist-tag.js#L29)).

File: commands/publish/index.js [hunk 9]
```diff
@@ -714,6 +720,16 @@ class PublishCommand extends Command {
 
     // undefined defaults to "latest" OR whatever is in pkg.publishConfig.tag
   }
+
+  getPreDistTag(pkg) {
+    if (!this.options.preDistTag) {
+      return;
+    }
+    const isPrerelease = (semver.prerelease(pkg.version) || []).shift();
```
Comment (evocateur) [inline commands/publish/index.js]: This logic is lifted from [here](https://github.com/lerna/lerna/blob/4e3f8d8087e0e2e8e83c7ea79558c517203583c2/core/package-graph/index.js#L25), I'd rather not duplicate it.
```suggestion
    const isPrerelease = this.packageGraph.get(pkg.name).prereleaseId;
```

File: commands/version/__tests__/version-command.test.js [hunk 10]
```diff
@@ -48,6 +48,9 @@ expect.addSnapshotSerializer(require("@lerna-test/serialize-git-sha"));
 
 describe("VersionCommand", () => {
   describe("normal mode", () => {
+    beforeEach(() => {
+      checkWorkingTree.mockReset();
```
Comment (evocateur) [inline commands/version/__tests__/version-command.test.js]: Why is this necessary? We are already [clearing mocks](https://github.com/lerna/lerna/blob/4e3f8d8087e0e2e8e83c7ea79558c517203583c2/jest.config.js#L4) after every test.

File: commands/version/__tests__/version-command.test.js [hunk 11]
```diff
@@ -134,6 +151,17 @@ describe("VersionCommand", () => {
...
 
+    it("does not throw if current ref is already tagged when using --force-publish", async () => {
+      checkWorkingTree.mockImplementationOnce(() => {
+        throw new Error("released");
+      });
+
+      const testDir = await initFixture("normal");
+      await lernaVersion(testDir)("--force-publish");
+
+      expect.assertions(0);
```
Comment (evocateur) [inline commands/version/__tests__/version-command.test.js]: I would prefer this take make an active assertion, such as `expect().not.toThrow()`. Also, unclear why this test is in this PR?

File: commands/version/index.js (hunk 7 above)
Comment (evocateur) [reply commands/version/index.js]: hrm, ok, this is confusing when conflated with the rest of the logic. I'd prefer this was in a separate PR for clarity.

File: commands/publish/index.js (hunk 8 above)
Comment (erquhart) [reply commands/publish/index.js]: Good catch, I wasn't understanding that the config snapshot should continue to show the `dist-tag` and ignore `pre-dist-tag` (as they can be used together).

File: commands/publish/index.js (hunk 9 above)
Comment (erquhart) [reply commands/publish/index.js]: Actually, we're working with a `Package`, not a `PackageGraphNode` - the version isn't up to date in the graph when we need to get the `preDistTag`, so have to determine `prereleaseId` from the `Package` instances in the mapper.

Considering that this logic was already duplicated [here](https://github.com/lerna/lerna/blob/4e3f8d8087e0e2e8e83c7ea79558c517203583c2/commands/version/index.js#L283), I'll just separate out the logic itself. It should probably be tested directly anyway, critical little bit of logic here.

File: commands/version/index.js (hunk 7 above)
Comment (erquhart) [reply commands/version/index.js]: This PR isn't complete without these four lines - graduate prereleases is facilitated by passing `--conventional-graduate`, and that fails without this fix. It just so happens that this same issue is also a known bug for `--force-publish`.

File: commands/version/__tests__/version-command.test.js (hunk 11 above)
Comment (erquhart) [reply commands/version/__tests__/version-command.test.js]: Regarding why this test is here, it's because this fixed [happened to be fixed](https://github.com/lerna/lerna/pull/1991/files/255aeb18db2035d663079be0f9daf9cc02b7df49#r271093980) here. Will address assertion style.

File: commands/version/__tests__/version-command.test.js (hunk 10 above)
Comment (erquhart) [reply commands/version/__tests__/version-command.test.js]: Resetting removes mock implementations (clearing doesn't). This shouldn't have been necessary since `mockImplementationOnce` is being used, a second look revealed that I shouldn't be mocking an implementation for an error that is **not** expected. Fixed.

File: commands/version/__tests__/version-command.test.js (hunk 11 above)
Comment (erquhart) [reply commands/version/__tests__/version-command.test.js]: Making sure this particular call *doesn't* throw is just a bad approach in general. Going to test for a positive outcome instead.

File: commands/publish/index.js (hunk 8 above)
Comment (erquhart) [reply commands/publish/index.js]: Fix: https://github.com/lerna/lerna/pull/1991/commits/850bc22e60d6fb7cac7c5543770a0cb3347bbf62

File: commands/publish/index.js (hunk 9 above)
Comment (erquhart) [reply commands/publish/index.js]: Fix: https://github.com/lerna/lerna/pull/1991/commits/70fb6301c660755c015a2c2795cd97272fdfd874

File: commands/version/__tests__/version-command.test.js (hunk 10 above)
Comment (erquhart) [reply commands/version/__tests__/version-command.test.js]: Fix: https://github.com/lerna/lerna/pull/1991/commits/86a88aa19807e3d25bf815bf6a35cfff1e278869

File: commands/version/__tests__/version-command.test.js (hunk 11 above)
Comment (erquhart) [reply commands/version/__tests__/version-command.test.js]: Fix: https://github.com/lerna/lerna/pull/1991/commits/a9ba5dfc2208f5d88a4d724a9394a84cba29435a

**Comment by erquhart** [comment]
@evocateur added commits to close the first round of review, replied to each of your actionable comments with commit links. I know the difficulty of reviewing a large changeset like this, especially from a new contributor. Hope it helps!

File: utils/prerelease-id-from-version/lib/prerelease-id-from-version.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+"use strict";
+
+const semver = require("semver");
+
+module.exports = prereleaseIdFromVersion;
+
+function prereleaseIdFromVersion(version) {
+  return (semver.prerelease(version) || []).shift();
+}
```
Comment (erquhart) [inline utils/prerelease-id-from-version/lib/prerelease-id-from-version.js]: Yes this feels a bit like overkill, but I really couldn't find a better path. This (ridiculously simple) logic is reused in different verticals across the codebase, like `core` and `command`, so it couldn't be dropped in an existing common lib. If you're up for adding a `lib` directory at the top level for little bits like this, I'm happy to drop this commit and add one that does that instead.

File: utils/collect-updates/lib/collect-packages.js (hunk 4 above)
Comment (evocateur) [reply utils/collect-updates/lib/collect-packages.js]: Awesome, thanks!

File: utils/collect-updates/lib/get-packages-for-option.js (hunk 5 above)
Comment (evocateur) [reply utils/collect-updates/lib/get-packages-for-option.js]: Good call

File: utils/prerelease-id-from-version/lib/prerelease-id-from-version.js (hunk 12 above)
Comment (evocateur) [reply utils/prerelease-id-from-version/lib/prerelease-id-from-version.js]: No worries, this is great!

File: package-lock.json [hunk 13]
```diff
@@ -14,36 +14,36 @@
       }
     },
     "@babel/core": {
-      "version": "7.2.2",
-      "resolved": "https://registry.npmjs.org/@babel/core/-/core-7.2.2.tgz",
-      "integrity": "sha512-59vB0RWt09cAct5EIe58+NzGP4TFSD3Bz//2/ELy3ZeTeKF6VTD1AXlH8BGGbCX0PuobZBsIzO7IAI9PH67eKw==",
+      "version": "7.3.4",
```
Comment (evocateur) [inline package-lock.json]: I'm not entirely sure why there are so many diffs in the lockfile, but if it isn't conflicting, I'm fine with it.

File: commands/publish/README.md [hunk 14]
```diff
@@ -154,6 +155,14 @@ lerna publish --canary --preid next
 When run with this flag, `lerna publish --canary` will increment `premajor`, `preminor`, `prepatch`, or `prerelease` semver
 bumps using the specified [prerelease identifier](http://semver.org/#spec-item-9).
 
+### `--pre-dist-tag <tag>`
```
Comment (evocateur) [inline commands/publish/README.md]: I don't see any of these flags documented in the yargs command builder for publish: https://github.com/lerna/lerna/blob/master/commands/publish/command.js

I suppose it's not vital, since yargs seems perfectly happy to consume them implicitly...

**Comment by evocateur** [comment]
That was a fun rebase (my own fault)

**Comment by erquhart** [comment]
Thank for getting it in!

**Comment by inside** [comment]
Hi,

I believe the behavior of `--conventional-commits` is now a breaking change and has to be used in conjunction with `--conventional-graduate` to keep the same behavior as before. Example:

```
% npx lerna -v 
3.13.4
% npx lerna version --no-push
...debug stripped...
Changes:
- @foo/bar: 1.0.0-rc.0 => 1.0.0
```

```
% npx lerna -v
3.14.1
% npx lerna version --no-push
...debug stripped...
Changes:
- @foo/bar: 1.0.0-rc.0 => 1.0.0-rc.1
```

If you guys can tell me if I'm missing something in the way semver works, thanks :+1:

**Comment by erquhart** [comment]
Conventional commits didn't really have any logical or intentional behavior for prereleases, so it's fair to say that using `--conventional-commits` on a prerelease was not supported prior to 3.14.0. If someone was using that flag against a prerelease before prerelease support was added, what resulted was unintentional on the part of Lerna. So if you happened to be depending on that unintentional functionality, it will feel like a breaking change, but the change didn't break any documented functionality of Lerna.

But to your immediate point, yes, to duplicate the result of `--conventional-commits` from before prerelease support was added, you'll need to add the `--conventional-graduate` flag.

**Comment by evocateur** [comment]
Wait, `--conventional-graduate` is necessary to preserve `1.0.0-rc.0` -> `1.0.0-rc.1`? That seems backwards, I would expect `--conventional-prerelease` to do that...

**Comment by erquhart** [comment]
No, `--conventional-commits` preserves the prerelease by itself. Adding `--conventional-prerelease` for a package that's already in prerelease doesn't do anything different. `--conventional-graduate` moves the package out of prerelease, which used to be the only behavior with the `--conventional-commits` flag.

**Comment by inside** [comment]
Thank you @erquhart it makes sense to me now! :smile:

**Comment by acao** [comment]
does anyone have this working in CI? I keep getting a prompts even with `--yes`. about to open a bug

**Comment by erquhart** [comment]
I've used in CI in the past without issue, been doing it manually the last few months, though. Maybe something changed?
